core/src/plugins/kubernetes/results-cache-fs.ts

```
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2018-2024 Garden Technologies, Inc. <info@garden.io>
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import type {
+  CacheableAction,
+  CacheableResult,
+  CacheKeyProvider,
+  ClearResultParams,
+  LoadResultParams,
+  ResultTrimmer,
+  ResultValidator,
+  StoreResultParams,
+} from "./results-cache.js"
+import { AbstractResultCache } from "./results-cache.js"
+import AsyncLock from "async-lock"
+import fsExtra from "fs-extra"
+import { join } from "path"
+import writeFileAtomic from "write-file-atomic"
+import { CACHE_DIR_NAME } from "../../constants.js"
+import type { Log } from "../../logger/log-entry.js"
+import { deline } from "../../util/string.js"
+import { renderZodError } from "../../config/zod.js"
+
+const { ensureDir, readFile, remove } = fsExtra
+
+/**
+ * Very simple implementation of file-system based cache
+ * to be used as a fallback storage for kubernetes Run and Test results.
+ *
+ * It uses cache keys to name and create JSON files,
+ * and stores the values in the JSON files in a configurable cache directory.
+ *
+ * All operations are fail-safe and no not throw any errors.
+ * All operations are concurrent-safe and protected by necessary locks.
+ *
+ * This class holds the minimal in-memory state,
+ * because each Run and Test result
+ * is usually read and written only once per Garden command execution.
+ */
+export class SimpleFileSystemCache<T> {
+  private readonly cacheDir: string
+  private readonly lock: AsyncLock
+
+  constructor(cacheDir: string) {
+    this.cacheDir = cacheDir
+    this.lock = new AsyncLock()
```

this lock protects against concurrent access within the same process, but users sometimes run multiple garden commands at the same time.

Should we use the SQLite database that we ship with anyway, and where we store tokens instead? SQLite is very robust against concurrent access even from multiple processes

---

Well, concurrent writes are possible if user run exactly the same command simultaneously from multiple terminals.

I'm not sure if this will be an often scenario. Afaik, we decided to get rid of SQLite in 0.13 because it slowed down the startup time of Garden CLI.

Now, if any issue occurs while the caching, it should not crash the Garden process. So, the current approach might be good enough for now. Let's try to find a better solution if any issues occur.

---

ah, I didn't know we removed sqlite (as the word still occurs in the code when looking it up)

I think that it's definitely possible that two different commands run simultaneously have actions in common that calculate to the same version hash, e.g. deploying to two different environments at the same time.

I think just ignoring errors is not quite enough here; E.g. if we're reading invalid data (e.g. that became invalid due to concurrent writes from 2 different processes), we need to at least remove the cache entry to allow for a valid entry to be saved next time.

We should also probably log a debug-level message when there is an error due to corrupt data.

I don't think we should just catch all and ignore, as is the case in the current PR: If something happens that is unrelated to data / schema validation, then we should throw that error and crash (to avoid programming mistakes to be overlooked)

---

Ah well, I had another look and noticed that you're using writeFileAtomic. That should prevent issues with concurrent writes.

So only the point about exception handling still stands 👍

---

I thought about it a little, and I believe if we're writing atomically we can completely eliminate the lock, I believe it is not necessary

---

Maybe it's necessary for remove on Windows, but we can also make sure that we never delete keys while we might receive writes (e.g. by deleting old keys during setup)

---

Fixed error handling in 698e7042a2c8ce928f1fd85e02993eec54a92fc7.

---

Yes, I also thought about the necessity of the lock. Concurrent reads are fine, and write happens atomically.

---

Delete is also fine on unix, as it's fine to delete a file and still read it or keep file descriptors open. only for delete on windows we need to ensure that we either handle the error, if another app has one of the files open, or we avoid deleting and reading simultaneously somehow.

---

Removed lock in 8a06ce426ef2741ddc1608651bb08671ecdffcd5.